Add new types for routing results

tahini opened the pull request.

tahini commented:
@greenscientist This PR is not complete, I want to add some unit tests, and the PR that goes above (actually using those types) does not work yet, so there may be a bug here, but I'd like a high level review for now. I'll add the PR that uses it tomorrow.

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,317 @@
+  class StepTransitRoutingResult : public StepRoutingResult {
+  public:
+    boost::uuids::uuid agencyUuid;
+    std::string agencyAcronym;
```
That's a lot of element here, kind of wonder if they should be regrouped in a subobject

tahini replied:
Maybe an optimization for later, but with only the tripUuid and nodeUuid, we can get pretty much all the rest, given that we have access to the data (data, which is now contained within the calculator object, so it is far from a given that access to the data is available where/when we need it). It could be an optimization for later though. Once this is in, we can open a issue to track when/if we want to change this.

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,317 @@
+    virtual T getResult() = 0;
+  };
+
+  class StepVisitorBase {
```
Not sure of it's role yet, but maybe a Visitor should be in its own file

greenscientist reviewed (changes requested):
It's missing class description, so we know what are the intents. Probably need to scather more comments around. The visitor class needs to be explained more

A couple of small structural things to adresse

tahini commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,317 @@
+    }
+  };
+
+  class NoRoutingFoundResult : public RoutingResultNew {
```
I wonder if we need the NoRouting result, or if it should just be a nullptr...

tahini replied:
Actually a nullptr object cannot be visited, so it's probably better to keep this class

greenscientist replied:
Do we need to convert a NoRouting result? Why would we need a visitor in this case?

tahini replied:
As explained in the other PR, NoRouting is boring for now, but it will contain a reason later. Unless we use try catch and throw the no routing explanation, but no routing is a valid workflow and using error throwing for it it not necessarily good practice.

Another option would be to include the NoRoutingFound use case in the SingleCalculationResult since it's when it will happen.

What do you think? 3 options:

* the one in this PR (different class)
* try-catch
* merge with SingleCalculation
* other?

tahini commented:
@greenscientist  I updated as per your comments: more documentation and split visitor/result and converter/converter_to_v1 to separate files. I'll wait for your review of this PR before updating the next one.

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,323 @@
+#include <boost/uuid/uuid.hpp>
+#include <stdlib.h>
+#include "point.hpp"
+#include "routing_result_visitor.hpp"
```
Could probably just be a forward declaration instead I think

greenscientist replied:
and this ^^

tahini replied:
This one no, we are calling methods from these classes.

greenscientist replied:
grumble. Not great to implement function in a .h, but ok.

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,323 @@
 #include <string>
+#include <boost/uuid/uuid.hpp>
+#include <stdlib.h>
+#include "point.hpp"
```
Could probably just be a forward declaration instead I think

greenscientist replied:
This ^^

tahini replied:
This one yes, can be forward declared

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,323 @@
+  /**
+   * @brief Base class for routing result steps. It accepts a visitor for the steps class hierarchy
+   */
+  class StepRoutingResult {
```
SHould this inherit from RoutingResultNew?

tahini replied:
No, steps are not results. It's really how it's explained.

greenscientist replied:
Still wondering about the naming. I don't feel strongly about it, but maybe just RoutingStep ? To differenciate with RoutingResult ?

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,323 @@
 {
-  
+  enum result_type { SINGLE_CALCULATION, NO_ROUTING_FOUND, ALTERNATIVES, ALL_NODES };
+  enum result_step_type { WALKING, BOARDING, UNBOARDING };
```
Shouldn't each of these type their own classes?

tahini replied:
They sort of are, but since we need to dynamically cast these enums are used to know which class to cast to. Is there a better way?

greenscientist reviewed (commented):
The class hierarchy seems to be lacking the top level. 
Am I missing something? Maybe it's lacking an high level doc/comment?

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,315 @@
+   * 
+   * TODO: Will be renamed to RoutingResult in later patch, when the struct is removed and this class is used instead
+   * 
+   * TODO: Even later, when function calls are detricated and split, there will be no use for a class hierarchy. Concrete result types
```
They could  still be in an hierarchy,  depending on the structure of the result. We'll need to see :)

greenscientist commented on include/routing_result.hpp:
```diff
@@ -3,12 +3,315 @@
+   * @brief A class describing a single accessible node details, with the time to reach the node
+   * and the number of transfers for the requested departure time
+   */
+  class AccessibleNodesResult {
```
Maybe also not end this in Result if not a RoutingResult ?

greenscientist commented on connection_scan_algorithm/include/result_to_v1.hpp:
```diff
@@ -0,0 +1,22 @@
+   */
+  class ResultToV1Response: public ResultToResponse {
+  public:
+    nlohmann::json resultToJsonString(RoutingResultNew& result, RouteParameters& params) override;
```
You dropped the virtual here

greenscientist replied:
Also, I would make them static. I don't see a need to instanciate a ResultToV1Response, we can just call the functions

tahini replied:
Don't I need to drop the virtual if I'm implementing it? Also, for static calls, you need to know exactly which object to call. What if the code path can support both v1 and v2? sending the converter as base type object allows to just call this method on it and get whatever api version we require.

greenscientist replied:
No, if you drop the virtual, you prohibit further inheritance. The différent is the absence of =0 which makes it virtual pure.

greenscientist replied:
For the static, would need to  analyse the code a little  bit more. Depending on how it's used.

greenscientist reviewed (commented):
Still a few small things

tahini commented on connection_scan_algorithm/include/result_to_response.hpp:
```diff
@@ -0,0 +1,21 @@
+
+#include "json.hpp"
+#include "routing_result.hpp"
+#include "parameters.hpp"
```
I guess I can forward declare these here too?

greenscientist replied:
Yeah probably. 
The json one is in a diffrent namespace, so might be annoying to forward declare, but the rest looks easy enough

greenscientist commented on connection_scan_algorithm/include/result_to_v1.hpp:
```diff
@@ -0,0 +1,23 @@
+   */
+  class ResultToV1Response: public ResultToResponse {
+  public:
+    virtual nlohmann::json resultToJsonString(RoutingResultNew& result, RouteParameters& params) override;
```
I guess override is a new keyword in C++!

greenscientist commented:
Almost there, a few forward declaration, and the discussion about the static function.
